Implement i18n

Mateusz-Dobrzynski opened the pull request:
Resolves #5:
- Introduces an i18n framework (`i18n-intl`),
- Updates dependencies,
- Internationalizes existing views,
- Extracts repetitions into separate components.

Mateusz-Dobrzynski committed:
feat: introduce i18n
- add dependencies
- configure i18n
- translate selected strings
- clean up code

Mateusz-Dobrzynski committed:
feat: make a language flag appear in a folded language switcher

Mateusz-Dobrzynski committed:
feat: internationalize login component

Mateusz-Dobrzynski committed:
refactor: merge dash_header and sidebar into a single file

Mateusz-Dobrzynski committed:
build!: change i18n framework to next-intl

shout-out to ejer

BREAKING CHANGE: upgrade next, drop i18next in favor of next-intl

Mateusz-Dobrzynski committed:
refactor: add missing translations, extract repetitions to components

jakubmanczak commented on src/i18n/language-utils.ts:
```diff
@@ -0,0 +1,14 @@
+    case "en":
+      return "English";
+    case "pl":
+      return "Polski";
```
```suggestion
      return "polski";
```

:P

Mateusz-Dobrzynski replied:
I remember us having this conversation a while back and your argument (correct me, if I'm wrong) that while lowercase "polski" is ugly, it's the standard. Since then, I've found more than enough counterpoints (e.g. Chrome and Firefox language settings) to support the decision of keeping it consistent. Notable apps use the version we like better anyway, so why force ourselves for the worse?

jakubmanczak replied:
My argument was that it's correct, in regards to language rules. And also used, more or less standard.

Mateusz-Dobrzynski replied:
If you were to use that word in a sentence, surely (_My first language is **English**_, but _Moim językiem ojczystym jest **polski**_). However, that's not the case here. In many contexts it's common to start words with capital letters, even though there are no grammatical rules enforcing that (unlike in e.g. German). We don't have to cling on to the rules, especially when we're not breaking any.

@julesity, what's your opinion on this? Context: should "Polski" should be lowercase here or can it be the way it is – and why?

<img width="161" height="100" alt="image" src="https://github.com/user-attachments/assets/175bc774-5e3b-4f7f-8ec6-f8cd7d7666ed" />

jakubmanczak commented on src/i18n/translations/pl.json:
```diff
@@ -0,0 +1,70 @@
+    }
+  },
+  "login": {
+    "msg_handle_empty": "Handle must not be empty.",
```
```suggestion
    "msg_handle_empty": "Nick nie może być pusty.",
```
String not translated. :P

Mateusz-Dobrzynski replied:
```suggestion
    "msg_handle_empty": "Nick nie może być pusty.",
```

jakubmanczak commented on src/i18n/translations/pl.json:
```diff
@@ -0,0 +1,70 @@
+    "msg_passwd_empty": "Hasło nie może być puste.",
+    "msg_invalid_credentials": "Błędne dane logowania",
+    "handle_label": "Nick",
+    "handle_aria_label": "Twój nick (nazwa użytkownika)",
```
```suggestion
    "handle_aria_label": "Twój nick",
```
Ah! So verbose!

jakubmanczak commented on src/i18n/translations/pl.json:
```diff
@@ -0,0 +1,70 @@
+      "log_out": "Wyloguj się"
+    },
+    "sidebar": {
+      "logo_by": "od",
```
😭

jakubmanczak commented on src/i18n/translations/pl.json:
```diff
@@ -0,0 +1,70 @@
+        "event_branding": "Szata graficzna"
+      },
+      "shareable": {
+        "catname": "materiały prasowe",
```
The press doesn't exist! Welcome to the new world order. But uhhh, I'd rethink this one translation probs. Such options are probably more relevant to social media more than anything, though I do realise prasa carries connotations and whatnots.

Mateusz-Dobrzynski replied:
```suggestion
        "catname": "media",
```

Mateusz-Dobrzynski replied:
```suggestion
        "catname": "media",
```

jakubmanczak commented on src/i18n/translations/pl.json:
```diff
@@ -0,0 +1,70 @@
+        "catname": "informacje o turnieju",
+        "competing_teams": "Drużyny",
+        "debate_motions": "Tezy debat",
+        "staff_judges_bias": "Organizatorzy i sędziowie",
```
Doesn't reflect bias. Curse the verbosity of my mother tongue.

Mateusz-Dobrzynski replied:
Does it have to, though? Where else in the app would you search for judges' affiliations? Besides, I'd question the usage of _bias_ in the English version to begin with, since it has strongly negative connotations (and there's nothing inherently wrong about being _affiliated_ with a judge, as long as they're not judging you) and does not reflect how it's named on the backend (and how I'm planning to name it in the rest of the app).

jakubmanczak replied:
I feel like there's still a benefit to signaling to organisers that there *could* be bias, even if it manifests subconsciously. Kind of like a reminder to take it seriously, since there's no guarantee tournaments actually will.

Mateusz-Dobrzynski replied:
Fair point. Changed it.

Mateusz-Dobrzynski committed:
fix: add missing translations, reduce verbosity

Co-authored-by: Jakub Mańczak <jakub@manczak.net>

Mateusz-Dobrzynski committed:
feat: localize metadata

jakubmanczak commented on src/app/[locale]/layout.tsx:
```diff
@@ -30,11 +32,13 @@ const FontLexend = Lexend({
-  description:
-    "Computer aided debate tournament organizing experience enrichment.",
-};
+export async function generateMetadata(): Promise<Metadata> {
+  const t = await getTranslations("metadata");
+  return {
+    title: t("title"),
+    description: t("description"),
+  } as Metadata;
+}
```
On the topic of metadata: if we don't separate or somehow distinguish polish and english versions in the URLs, changing metadata like that will likely hurt SEO. We should consider how we'd do this before moving forward with dynamic metadata.

Mateusz-Dobrzynski replied:
Fixed. Implemented routing-based localization (e.g. `/en/login` will return a website in English).

jakubmanczak commented on src/components/icons/FlagIcon.tsx:
```diff
@@ -0,0 +1,22 @@
+      flagCode = props.language;
+    }
+  }
+  return <span className={"h-[12px] scale-x-[1.15] fi fi-" + flagCode}></span>;
```
I'd consider switching to SVGs before mixing tailwind with npm-imported css. Your call.

```suggestion
  return <span className={`h-[12px] scale-x-[1.15] fi fi-${flagCode}`}></span>;
```

jakubmanczak commented on src/components/settings/LanguageSwitcher.tsx:
```diff
@@ -0,0 +1,33 @@
+import GenericSelect from "../ui/GenericSelect";
+import { LanguageCode } from "@/types/Language";
+import { FlagIcon } from "../icons/FlagIcon";
+import { getEndonym, LANGUAGE_COOKIE, languages } from "@/i18n/language-utils";
```
As far as I'm aware, while aren't required to get consent for 'strictly necessary' cookies like language or color schemes under GDPR, we do need to disclose the details of their use somewhere. A thing to keep in mind - maybe a subpage about it somewhere?

Mateusz-Dobrzynski replied:
By default, next-intl uses session cookies to be "[compliant out of the box](https://next-intl.dev/docs/routing/configuration#locale-cookie-gdpr)". I'm leaving it that way until we come up with a sensible idea to present cookie disclaimers.

Mateusz-Dobrzynski committed:
feat: implement language-based URLs

Mateusz-Dobrzynski committed:
fix: i18n-related fixes

- make URL accurately reflect website language
- in case of locale-less URLs, redirect to localized pages instead of returning 404

Mateusz-Dobrzynski committed:
refactor: code cleanup

jakubmanczak commented on src/i18n/translations/pl.json:
```diff
@@ -0,0 +1,74 @@
+    "msg_invalid_credentials": "Błędne dane logowania",
+    "handle_label": "Nick",
+    "handle_aria_label": "Twój nick",
+    "handle_placeholder": "nick",
```
This will make it inline with the password placeholder semantically and visually, as just 'nick' is disturbingly short in comparison.
```suggestion
    "handle_placeholder": "twój_nick",
```

jakubmanczak commented on src/i18n/translations/pl.json:
```diff
@@ -0,0 +1,74 @@
+    "login_redirect_footer": "Możesz też wrócić do <login>strony logowania</login>."
+  },
+  "debatecore": {
+    "logo_subtitle": "Zaplanuj swój turniej debat",
```
The translation causes the subtitle to be misaligned (not aligned to the left, to be flush with the letter g's descender); this misalignment should either be fixed or the translation removed/reverted until the misalignment is fixed. Minor nit: the translation being shorter and not extending beyond the logo also irks me. Can be ignored probably.

<img width="308" height="112" alt="Image" src="https://github.com/user-attachments/assets/29a12360-0df3-44ba-a14e-a8e59dc3da65" />

<img width="308" height="112" alt="Image" src="https://github.com/user-attachments/assets/723e12ad-896f-4692-abc2-cd30736b0c02" />

```suggestion
    "logo_subtitle": "The Debate Tournament Planner",
```

jakubmanczak reviewed (changes requested):
I can't seem to be able to log in as initial infrastructure admin. Can you? If yes, we should schedule a chat to sort out the differences in our approaches, because one of us is doing something wrong.

Also featuring: miscellaneous string changes.

Mateusz-Dobrzynski committed:
refactor: improve polish translations per code review suggestions

jakubmanczak commented:
Compilation is broken due to eslint errors on unused variables. Might be nice to add an ignore clause to the eslint config. Offhand comment: what version npm are you using, that you're not getting this?

```esling.config.mjs
rules: {
      "@next/next/no-img-element": "off",
      "@typescript-eslint/no-unused-vars": "off",
    },
```

Otherwise, it complains about malformed JSON responses. Again, we must be doing something very differently; it's really worrying that I'm unable to even handtest the frontend while you're presumably unaffected.

Mateusz-Dobrzynski committed:
build: change eslint config

Mateusz-Dobrzynski commented:
Changes to eslint config were made.

>What version npm are you using, that you're not getting this?

Well, I was. However, unused variables were already present in the code, so I assumed it's fine by our standards and ignored it (maybe if I remembered to build the app, would've done it differently).

>Otherwise, it complains about malformed JSON responses.

Unclear, not sure what to do about it. I double-checked functionality of the app using [0713216](https://github.com/debatecore/tau/pull/74/commits/0713216f43900474dfe421e83130faa5a0657e46) as backend. The fix introduced in https://github.com/debatecore/tau/pull/74 is necessary, it won't work with what's present master branch (otherwise I wouldn't have made the fix to begin with).

jakubmanczak commented:
As it turns out, the malformed JSON response was due to another bug that only occurs when no tournaments are present in the database - an empty response is sent instead of an empty json array. (Clearing your `tau_dbdevdata` docker volume when doublechecking such things is expected as it lands you a clean starting point!). This being said, this is an issue with another part of the codebase, making no difference here.

jakubmanczak commented on src/i18n/translations/pl.json:
```diff
@@ -0,0 +1,73 @@
+        "catname": "informacje o turnieju",
+        "competing_teams": "Drużyny",
+        "debate_motions": "Tezy debat",
+        "staff_judges_bias": "Organizatorzy, sędziowie i powiązania",
```
I know the initiative to reflect bias came from me, but a direct translation is way too long and looks comically off in the menu itself. Perhaps this would be more compact while keeping with the idea.
```suggestion
        "staff_judges_bias": "Personel i powiązania",
```

Mateusz-Dobrzynski replied:
Yeah, that's a good idea.

Mateusz-Dobrzynski committed:
i18n: simplify polish translation of "stuff and judges"

Co-authored-by: Jakub Mańczak <jakub@manczak.net>

jakubmanczak commented:
<img width="425" height="348" alt="image" src="https://github.com/user-attachments/assets/7e0bab7d-6b30-485b-82fe-8fce630f36a3" />
